File: python/railostools/ttb/string.py
import typing

import railostools.ttb.components as ros_comp


def concat(*args, join_type=None) -> str:
    if not args:
        return ""
    if (join_type is ros_comp.Element) or isinstance(args[0], ros_comp.Element):
        return "\0".join(str(i) for i in args)
    else:
        return ";".join(str(i) for i in args)


def split(
    component_str: str, split_type=None
) -> typing.Union[typing.Tuple[typing.List[str], ...], typing.List[str]]:
    if split_type is ros_comp.Element:
        _split_char = "\0"
    elif split_type is ros_comp.Service:
        _split_char = ","
    else:
        _split_char = ";"
    return component_str.split(_split_char)

File: cpp/src/ttb/string.cxx
#include "railostools/ttb/string.hxx"
#include <iostream>

std::string RailOSTools::concat(const std::vector<std::string>& components, const ComponentType& join_type) {
    if(components.empty()) return "";

    std::string out_str_;
    std::string delimiter_;

    switch(join_type) {
        case(ComponentType::Element):
            delimiter_ = "\0"s;
            break;
        case(ComponentType::Service):
            delimiter_ = ",";
            break;
        default:
            delimiter_ = ";";
    }

    for(int i{0}; i < components.size(); ++i) {
        out_str_ += components[i];
        if(i < components.size() - 1)  out_str_ += delimiter_;
    }

    return out_str_;
}

std::vector<std::string> RailOSTools::split(const std::string& component_str, const ComponentType& split_type) {
    
    std::vector<std::string> components_;
    std::string delimiter_;

    switch(split_type) {
        case(ComponentType::Element):
            delimiter_ = "\0"s;
            break;
        case(ComponentType::Service):
            delimiter_ = ",";
            break;
        default:
            delimiter_ = ";";
    }

    std::string::size_type component_start_ = component_str.find_first_not_of(delimiter_, 0);
    std::string::size_type component_end_ = component_str.find_first_of(delimiter_, component_start_);

    while(std::string::npos != component_start_ || std::string::npos != component_end_) {
        components_.push_back(component_str.substr(component_start_, component_end_ - component_start_));
        component_start_ = component_str.find_first_not_of(delimiter_, component_end_);
        component_end_ = component_str.find_first_of(delimiter_, component_start_);
    }

    return components_;
}